### PR TITLE
Refactor group profile to include before hooks

### LINKED ISSUE #1878: "Top N Slowest Example Group" from `--profile` does not account for time spent in `before(:context)` hooks
The implementation calculates the total example group time by summing the time of each individual example.  Problem is, the time in before/after `:context` hooks is not included in individual example times so it's completely ignored.  This should be fixed, because if you are using `:context` hooks it may be because of a slow operation and you wouldn't want it ignored when trying to find the slowest example groups.

### PR DESCRIPTION
**Pull request by JonRowe**
This is #1942 refactored, fixes #1878 ping @myronmarston

### CODE REVIEW

File: lib/rspec/core/configuration.rb [hunk 1]
```diff
@@ -828,6 +828,18 @@ def profile_examples
       end
 
       # @private
+      def profiler
+        @profiler ||=
+          begin
+            if value_for(:profile_examples) { @profile_examples }
```
Comment (myronmarston) [inline lib/rspec/core/configuration.rb]: Can't this use the `profile_examples` method instead of repeating its logic to get a value?

File: lib/rspec/core/configuration.rb [hunk 2]
```diff
@@ -828,6 +828,18 @@ def profile_examples
...
+              Profiler.new
+            else
+              NoProfiler.new
+            end
+          end
+      end
```
Comment (myronmarston) [inline lib/rspec/core/configuration.rb]: There's a bug here...consider what would happen here:

``` ruby
config.profile_examples = true
config.profiler
config.profile_examples = false
config.profiler
```

Or:

``` ruby
config.profile_examples = false
config.profiler
config.profile_examples = true
config.profiler
```

In either case, the value of `config.profiler` does not update to reflect the mutated value of `profile_examples`.  Our normal way to deal with this is to set `@profiler = nil` in `profile_examples=` so that it clears it out if it changes.

File: lib/rspec/core/formatters/json_formatter.rb [hunk 3]
```diff
@@ -60,8 +60,7 @@ def dump_profile(profile)
         # @api private
         def dump_profile_slowest_examples(profile)
           @output_hash[:profile] = {}
-          sorted_examples = profile.slowest_examples
-          @output_hash[:profile][:examples] = sorted_examples.map do |example|
+          @output_hash[:profile][:examples] = profile.slowest_examples.map do |example|
```
Comment (myronmarston) [inline lib/rspec/core/formatters/json_formatter.rb]: Is this change related to this PR or just a refactoring?

File: lib/rspec/core/profiler.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
+module RSpec
+  module Core
+    # @private
+    class NoProfiler
+      def example_groups
+        {}
+      end
+    end
```
Comment (myronmarston) [inline lib/rspec/core/profiler.rb]: This class is stateless.  Is it worth making it a singleton module and then using `NoProfiler` directly instead of `NoProfiler.new` to avoid the extra object allocation?

File: lib/rspec/core/profiler.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    # @private
+    class Profiler
+      Formatters.register self, :example_group_started, :example_group_finished, :example_started
```
Comment (myronmarston) [inline lib/rspec/core/profiler.rb]: We already have a profile formatter:

https://github.com/rspec/rspec-core/blob/master/lib/rspec/core/formatters/profile_formatter.rb

Why have 2 profile formatters?  Seems like this stuff could be merged into the existing formatter.

File: lib/rspec/core/configuration.rb (hunk 2 above)
Comment (myronmarston) [inline lib/rspec/core/configuration.rb]: Actually, why do we even need a profile formatter exposed off of `config`?

File: lib/rspec/core/notifications.rb [hunk 6]
```diff
@@ -496,7 +496,8 @@ def duplicate_rerun_locations
     # @attr duration [Float] the time taken (in seconds) to run the suite
     # @attr examples [Array<RSpec::Core::Example>] the examples run
     # @attr number_of_examples [Fixnum] the number of examples to profile
-    ProfileNotification = Struct.new(:duration, :examples, :number_of_examples)
+    # @attr profiler [RSpec::Core::Profiler] the profiler used to capture examples
+    ProfileNotification = Struct.new(:duration, :examples, :number_of_examples, :profiler)
```
Comment (myronmarston) [inline lib/rspec/core/notifications.rb]: I don't think the profiler formatter needs to be exposed of  profile notification.  Seems like once you merge the example/example group event handling into the existing profiler formatter this will no longer be needed on the notification object anymore, right?

File: lib/rspec/core/formatters/json_formatter.rb (hunk 3 above)
Comment (JonRowe) [inline lib/rspec/core/formatters/json_formatter.rb]: Well we're not sorting the examples here so why assign a local variable?

File: lib/rspec/core/configuration.rb (hunk 2 above)
Comment (JonRowe) [inline lib/rspec/core/configuration.rb]: It's not a formatter, it's actually doing the profiling.

File: lib/rspec/core/notifications.rb (hunk 6 above)
Comment (JonRowe) [inline lib/rspec/core/notifications.rb]: The notification object needs the profiler, it doesn't have to be exposed publicly I suppose...

File: lib/rspec/core/profiler.rb (hunk 4 above)
Comment (JonRowe) [inline lib/rspec/core/profiler.rb]: I suppose, especially if we end up not memoizing it.

File: lib/rspec/core/profiler.rb (hunk 5 above)
Comment (JonRowe) [inline lib/rspec/core/profiler.rb]: It's not a formatter, it's just using the callback system.

Comment (JonRowe) [inline lib/rspec/core/profiler.rb]: Further it can't be merged because the logic is used by 2 formatters, this is removing repeated logic.

Comment (myronmarston) [inline lib/rspec/core/profiler.rb]: > It's not a formatter, it's just using the callback system.

Well, it's registered as a formatter, and I thought that was what your definition of a formatter was...given that I've previously named a class `ExceptionFormatter` and you asked me to rename since it wasn't registered as a formatter.

Regardless, why can't this callback handling be moved into the existing `ProfilerFormatter`?  Seems like it would have full access to the data tracked by the callbacks without having to complicate config and/or notifications with it.

You say:

> Further it can't be merged because the logic is used by 2 formatters, this is removing repeated logic.

I don't understand.  Which 2 formatters are using this?  It's brand new anyway...  I really don't understand why this can't be merged into the other one.

Comment (JonRowe) [inline lib/rspec/core/profiler.rb]: > Well, it's registered as a formatter, and I thought that was what your definition of a formatter was...given that I've previously named a class ExceptionFormatter and you asked me to rename since it wasn't registered as a formatter.

It's abusing the `FormatterLoader`, but you can't actually load this as formatter, I can specify the notifications another way if you'd prefer. I asked you to rename it because it *wasn't* a formatter, not because it didn't use the lo

> Regardless, why can't this callback handling be moved into the existing ProfilerFormatter? Seems like it would have full access to the data tracked by the callbacks without having to complicate config and/or notifications with it.

Because it's used in the `JSONFormatter` too, and the idea is to make it accesible to any formatter that wishes to have profile output.

> I don't understand. Which 2 formatters are using this? It's brand new anyway... I really don't understand why this can't be merged into the other one.

The logic isn't brand new, it's extrapolated from `ProfileNotification` as per the original PR, the values are used in both the `ProfileFormatter` (for command line output) and `JSONFormatter`. Ostensibly it could be used by any formatter.

File: lib/rspec/core/configuration.rb (hunk 2 above)
Comment (JonRowe) [inline lib/rspec/core/configuration.rb]: > There's a bug here...consider what would happen here:

> ``` ruby
config.profile_examples = true
config.profiler
config.profile_examples = false
config.profiler
```

> Or:

> ``` ruby
config.profile_examples = false
config.profiler
config.profile_examples = true
config.profiler
```

> In either case, the value of `config.profiler` does not update to reflect the mutated value of `profile_examples`.  Our normal way to deal with this is to set `@profiler = nil` in `profile_examples=` so that it clears it out if it changes.

Good point, to enable this we'll have to profile all the time making the config profiler irrelevant.

**Comment by JonRowe** [comment]
I updated this slightly

File: lib/rspec/core.rb [hunk 7]
```diff
@@ -31,6 +31,7 @@
   option_parser
   configuration_options
   runner
+  profiler
```
Comment (myronmarston) [inline lib/rspec/core.rb]: Can we avoid the cost of loading this file if the user isn't using `--profile`?

File: lib/rspec/core/formatters/json_formatter.rb (hunk 3 above)
Comment (myronmarston) [inline lib/rspec/core/formatters/json_formatter.rb]: It's fine.  I was just making sure there wasn't some behavioral change here I wasn't noticing.

File: lib/rspec/core.rb (hunk 7 above)
Comment (JonRowe) [inline lib/rspec/core.rb]: Originally we had both profilers in that file, which would allow a cost reduction when not using `--profile` but in order to make `config#profile_examples` dynamic we must always have the profiler on (it's not actually doing much just capturing stats that aren't captured during run time normally)

File: lib/rspec/core/formatters/json_formatter.rb (hunk 3 above)
Comment (JonRowe) [inline lib/rspec/core/formatters/json_formatter.rb]: Nope, just clean up :)

File: lib/rspec/core/notifications.rb (hunk 6 above)
Comment (myronmarston) [inline lib/rspec/core/notifications.rb]: So what's weird to me about putting the profiler object on here is that registered formatter objects receive notification objects in the various event method calls.  Here you've making a notification object which holds a reference to a formatter.  So you'll wind up passing a notification containing a formatter to another formatter...which seems weird, circular and confusing.

Why would a formatter have to gain access to another formatter via a notification?

I think the notification just needs to have the profiling result data, and not the formatter object itself, right?  Given that...would it be a better separation of concerns if the notification got a reference to the profiling results but not the profiler itself?

Comment (JonRowe) [inline lib/rspec/core/notifications.rb]: > Why would a formatter have to gain access to another formatter via a notification?

It's not a formatter, it's the profiler, `RSpec::Core::Profiler` not `RSpec::Core::ProfileFormatter`. This is *given* to `RSpec::Core::ProfileFormatter`. One way, no circular dependency.

Comment (JonRowe) [inline lib/rspec/core/notifications.rb]: Actually I suppose I could pass in the example groups...

Comment (JonRowe) [inline lib/rspec/core/notifications.rb]: Refactored to do that

File: lib/rspec/core.rb (hunk 7 above)
Comment (myronmarston) [inline lib/rspec/core.rb]: > but in order to make `config#profile_examples` dynamic we must always have the profiler on

A big reason tracking example group execution time in a registered formatter (even if you don't call it one but I do!) appealed to me is that we want the added runtime cost of tracking times for profiling to only be paid if the user has opted in via `--profile`.  It makes me a sad panda to always pay the cost of tracking that when the user hasn't turned on profiling.

IMO, we don't need to support dynamic changes to `config#profile_examples` after the suite has started running.  We can treat it like adding a formatter: it has to be done before the suite starts to take affect.  (A user obviously can't expect to turn on profiling half way through the suite run and expect it to work).

My problem with what you had before is that `config` exposed the `profile_examples` option and a derived attribute, and it seems bad for an object to expose a derived attribute but then not keep it in sync with the source value.  Now that you're not exposing the profiler object off of `config`, I'm not concerned about that anymore.  Now I just view `profile_examples` as a boolean toggle that is checked once, when setting things up to start running the suite, and ignored the rest of the time.  Users can toggle it back and forth all they want before the suite starts and it'll work fine still.

In effect, I think we should only load the profiler and register it with the reporter if we get [here](https://github.com/rspec/rspec-core/blob/5375dcdfdb5460a51d5003de76eb124997fa26ea/lib/rspec/core/formatters.rb#L126).  That'll avoid the cost of loading and using it if the user hasn't opted-in to it.

File: lib/rspec/core/profiler.rb [hunk 8]
```diff
@@ -0,0 +1,29 @@
+module RSpec
+  module Core
+    # @private
+    class Profiler
+      NOTIFICATIONS = [:example_group_started, :example_group_finished, :example_started]
+
+      def initialize
+        @example_groups = Hash.new { |h, k| h[k] = Hash.new(0) }
```
Comment (myronmarston) [inline lib/rspec/core/profiler.rb]: This line confuses me a bit, because it suggests that all the nested hashes should have incrementing integer values...but `:start`, `:description` and `:total_time` are not integers.  WDYT of this instead?

```` ruby
@example_groups = Hash.new { |h, k| h[k] = { :count => 0 } }
```

That provides the count default of `0` (so you can simply increment it below w/o worrying about if it's the first time or not) without confusingly suggesting all the entries in the nested hash are integers.

File: lib/rspec/core/reporter.rb [hunk 9]
```diff
@@ -18,6 +18,8 @@ def initialize(configuration)
       @failed_examples = []
       @pending_examples = []
       @duration = @start = @load_time = nil
+      @profiler = Profiler.new
```
Comment (myronmarston) [inline lib/rspec/core/reporter.rb]: :+1: I much prefer having this owned by the reporter rather than config.

File: spec/rspec/core/profiler_spec.rb [hunk 10]
```diff
@@ -0,0 +1,70 @@
+require 'rspec/core/profiler'
+
+RSpec.describe 'RSpec::Core::Profiler' do
+  let(:profiler) { RSpec::Core::Profiler.new }
+
+  it 'has starts with an empty hash of example_groups' do
```
Comment (myronmarston) [inline spec/rspec/core/profiler_spec.rb]: What does "has starts" mean?

File: lib/rspec/core/profiler.rb [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+      end
+
+      attr_reader :example_groups
+
+      def example_group_started(notification)
+        @example_groups[notification.group.id][:start] = Time.now
+        @example_groups[notification.group.id][:description] = notification.group.top_level_description
```
Comment (myronmarston) [inline lib/rspec/core/profiler.rb]: Do you think there's an advantage to using the `id` as the hash key instead of the group itself?  The group is a class, which can be a hash key just fine, so the `id` seems like an unnecessary layer of indirection.  Plus there's some computation that has to be done to generate it so we can avoid that by using the `group` as the key.

File: spec/rspec/core/profiler_spec.rb [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+        end
+    end
+
+    describe '#example_group_started' do
+      it 'records example groups start time and description via id' do
+        expect {
+          profiler.example_group_started group_notification group
+        }.to change { profiler.example_groups[id] }.to include(
+          :start => now, :description => description
+        )
```
Comment (myronmarston) [inline spec/rspec/core/profiler_spec.rb]: These `change { }.to include` expectations are prone to the issue raised in rspec/rspec-expectations#617: it'll pass as long as the hash changes somehow and includes these entries at the end, even if it already included these entries at the start.  I think it would be better to do something like:

``` ruby
to change { }.
  from(a_hash_excluding(:start, :description).
  to(a_hash_including(:start => now, :description => description)
```

That way it's explicit and avoids that issue.

File: spec/rspec/core/profiler_spec.rb [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+  it 'has starts with an empty hash of example_groups' do
+    expect(profiler.example_groups).to be_empty.and be_a Hash
+  end
+
+  context 'when hooked into the reporter' do
+    include FormatterSupport
+
+    let(:id) { 42 }
+    let(:description ) { "My Group" }
+    let(:now) { ::Time.now }
```
Comment (myronmarston) [inline spec/rspec/core/profiler_spec.rb]: In my experience, when you don't care what the time used is it's best to specify it as an explicit static value (e.g. `Time.utc(2015, 5, 28, 12, 30)`).  Otherwise you're prone to having failures when leap seconds or DST transitions occur while running the suite.

**Comment by myronmarston** [comment]
Thanks, @JonRowe, this is looking good.

File: lib/rspec/core/profiler.rb [hunk 14]
```diff
@@ -0,0 +1,29 @@
+module RSpec
+  module Core
+    # @private
+    class Profiler
```
Comment (myronmarston) [inline lib/rspec/core/profiler.rb]: Given that this really only provides profiling for example groups (since each individual example profiles itself), do you think it should be called `ExampleGroupProfiler`?  That would also help differentiate it from the `ProfileFormatter`.

On a side note, I think I'd prefer to see this in the `formatters` directory.  No, it's not an object that formats output, but it gets registered as a formatter with the reporter and that's where we generally put those classes.  Plus it's nice not to increase the file count in the already crowded `rspec/core` dir.

File: lib/rspec/core/profiler.rb (hunk 5 above)
Comment (myronmarston) [inline lib/rspec/core/profiler.rb]: Thanks for explaining this.  I forgot about the `JSONFormatter`...

File: lib/rspec/core.rb (hunk 7 above)
Comment (JonRowe) [inline lib/rspec/core.rb]: >A big reason tracking example group execution time in a registered formatter (even if you don't call it one but I do!) 

In our terms a formatter is responsible for output to the user, in some cases they format the data themselves in others they use the presenter pattern to share workload between them, the Profiler does neither of these, it neither outputs anything or formats anything, it **just** captures values. It's no more a formatter than the reporter, or the runner or any other core component of RSpec.

The reporter follows (roughly) the event bus pattern, it sends out events when triggered to any registered **listener** of which the primary example are formatters I will admit, but this is not a formatter. In the same way your exception presenter wasn't a formatter. Naming is important.

> appealed to me is that we want the added runtime cost of tracking times for profiling to only be paid if the user has opted in via --profile. It makes me a sad panda to always pay the cost of tracking that when the user hasn't turned on profiling.

> IMO, we don't need to support dynamic changes to config#profile_examples after the suite has started running. We can treat it like adding a formatter: it has to be done before the suite starts to take affect. (A user obviously can't expect to turn on profiling half way through the suite run and expect it to work).

Thats fine but theres probably edge cases involved and I'd rather go back to setting it at the start, off or on, no half way ground (because otherwise the profiler has to go back to config, or we have to setup two way communication between the config and the reporter (to turn on and off the profiler), which I'd rather not do.

> My problem with what you had before is that config exposed the profile_examples option and a derived attribute, and it seems bad for an object to expose a derived attribute but then not keep it in sync with the source value. Now that you're not exposing the profiler object off of config, I'm not concerned about that anymore. Now I just view profile_examples as a boolean toggle that is checked once, when setting things up to start running the suite, and ignored the rest of the time. Users can toggle it back and forth all they want before the suite starts and it'll work fine still.

They won't be able to do that if the Profiler / NoProfiler is initialised in the reporter unless we somehow hook into the repoter from the config, I'd rather we set it once.

> In effect, I think we should only load the profiler and register it with the reporter if we get here. That'll avoid the cost of loading and using it if the user hasn't opted-in to it.

File: lib/rspec/core/profiler.rb (hunk 14 above)
Comment (JonRowe) [inline lib/rspec/core/profiler.rb]: I'm strongly against polluting the formatters namespace with more things that are not formatters. The reporter is not in the formatters namespace ergo this isn't.

File: lib/rspec/core.rb (hunk 7 above)
Comment (myronmarston) [inline lib/rspec/core.rb]: > In our terms a formatter is responsible for output to the user, in some cases they format the data themselves in others they use the presenter pattern to share workload between them, the Profiler does neither of these, it neither outputs anything or formats anything, it just captures values. It's no more a formatter than the reporter, or the runner or any other core component of RSpec.

> The reporter follows (roughly) the event bus pattern, it sends out events when triggered to any registered listener of which the primary example are formatters I will admit, but this is not a formatter. In the same way your exception presenter wasn't a formatter. Naming is important.

What you're saying makes sense, but reveals that I totally misunderstood your request on the other PR to rename the thing form `ExceptionFormatter` to something else.  You're right that the things we register with the reporter are "listeners" (not necessarily output formatters) but unfortunately, our public API speaks only of formatters.  The public API is `add_formatter`, not `add_listener`.  It might be worth changing that in RSpec 4 (since registering a listener is the generic way to be notified of events...and it's always bothered me that a user has to register a "formatter" for that...) to `add_listener`.  Although, that might not be worth the pain it could cause for users...who knows.

I've been thinking of the "formatters" directory as the place where all the listener class definitions are, but we just happen to keep calling them "formatters" since that's what they've been called for years.  It feels really odd to me to split where the listener classes are defined (since they feel natural to be grouped to me) but you're right that it's also odd to put a non-formatting listener in `formatters`.  I wish the directory was called "listeners", frankly :(.

Anyhow, I think we've been talking past each other a bit and our different usage of "formatter" is largely why, I think.

> Thats fine but theres probably edge cases involved and I'd rather go back to setting it at the start, off or on, no half way ground (because otherwise the profiler has to go back to config, or we have to setup two way communication between the config and the reporter (to turn on and off the profiler), which I'd rather not do.

We've been treating the `profile_examples` option in the way I'm suggesting for all RSpec 3 release -- after all, the [ProfileFormatter is only registered if `profile_examples` is set when `setup_default` is executing](https://github.com/rspec/rspec-core/blob/5375dcdfdb5460a51d5003de76eb124997fa26ea/lib/rspec/core/formatters.rb#L124-L126).  If the `ProfileFormatter` is registered, it's essential that this new listener is also registered, right?  Otherwise the `ProfileFormatter` won't work properly.  On the flip side, if the `ProfileFormatter` isn't registered, then this new listener serves no purpose, and it's wasted computation to register it and have it listen to the events.

So why not register them both together at the same time based on the value of the config option at that point?

I don't understand what the two way communication issue you refer to is or why this would necessitate moving it back onto the config object (where it doesn't belong, IMO).

One thing that might help is to consider having the reporter always initialize (and expose) a `profile_results` hash.  When instantiating the Profile listener we can then pass that to it so that it can populate it as it listens.  That way, the reporter always has a valid `profile_results` object that it can provide to the notification regardless of whether or not the profiler listener was actually used.  It's just that it'll be empty if the profile listener wasn't used, which makes sense, IMO.

> They won't be able to do that if the Profiler / NoProfiler is initialised in the reporter unless we somehow hook into the repoter from the config, I'd rather we set it once.

Not following you here again.  I thought we did away with `NoProfiler`?  Why does this necessitate bringing that back?

Sorry if it seems like I'm being argumentative here...I'm not intending to be.  I just care a lot about organizing RSpec's code in such a way that for optional features like the profiling, users only pay the runtime cost of that extra computation if they've opted-in to use the feature (kinda like how you care so much about stdlib usage!).  I think such an approach is essential to keeping RSpec fast even as we add and refine features.

Maybe what I'm asking for isn't easily possible for a reason I'm not yet seeing.  But I'd like to at least understand why if that's the case.

Comment (JonRowe) [inline lib/rspec/core.rb]: > What you're saying makes sense, but reveals that I totally misunderstood your request on the other PR to rename the thing form `ExceptionFormatter` to something else.  You're right that the things we register with the reporter are "listeners" (not necessarily output formatters) but unfortunately, our public API speaks only of formatters. The public API is `add_formatter`, not `add_listener`.  It might be worth changing that in RSpec 4 (since registering a listener is the generic way to be notified of events...and it's always bothered me that a user has to register a "formatter" for that...) to `add_listener`.  Although, that might not be worth the pain it could cause for users...who knows.

The reporter does have `add_listener`, `add_formatter` does something else, it sets up output streams, instantiates a formatter and then registers it as a listener on the reporter. It's possible already for extensions to utilise that hook without going through the process of registering a formatter, the reason we do that for formatters is they need a way to be distinguished from legacy formatters and it makes it easier to specify them from the command line etc.


> I've been thinking of the "formatters" directory as the place where all the listener class definitions are, but we just happen to keep calling them "formatters" since that's what they've been called for years.  It feels really odd to me to split where the listener classes are defined (since they feel natural to be grouped to me) but you're right that it's also odd to put a non-formatting listener in `formatters`.  I wish the directory was called "listeners", frankly :(.

I don't, it'd be ok with having two directories though.

> We've been treating the `profile_examples` option in the way I'm suggesting for all RSpec 3 release -- after all, the [ProfileFormatter is only registered if `profile_examples` is set when `setup_default` is executing](https://github.com/rspec/rspec-core/blob/5375dcdfdb5460a51d5003de76eb124997fa26ea/lib/rspec/core/formatters.rb#L124-L126).  If the `ProfileFormatter` is registered, it's essential that this new listener is also registered, right?  Otherwise the `ProfileFormatter` won't work properly.  On the flip side, if the `ProfileFormatter` isn't registered, then this new listener serves no purpose, and it's wasted computation to register it and have it listen to the events.

> So why not register them both together at the same time based on the value of the config option at that point?

This I could do.

> I don't understand what the two way communication issue you refer to is or why this would necessitate moving it back onto the config object (where it doesn't belong, IMO).

Theres no two way communication required as long as you're ok with:

 ```Ruby
config.profile_examples = true
config.profile_examples = false
```

Not actually removing the profiler or the profile formatter.

> One thing that might help is to consider having the reporter always initialize (and expose) a `profile_results` hash.  When instantiating the Profile listener we can then pass that to it so that it can populate it as it listens.  That way, the reporter always has a valid `profile_results` object that it can provide to the notification regardless of whether or not the profiler listener was actually used.  It's just that it'll be empty if the profile listener wasn't used, which makes sense, IMO.

Thats what NoProfiler was for, I didn't want to conflate the reporter with responsibility it didn't need.

> Not following you here again.  I thought we did away with `NoProfiler`?  Why does this necessitate bringing that back?

It's a null object version of the profiler... if you have a way to turn profiling off you need one or you have to add if statements or unrelated state to the reporter, both of which I'd rather not do.

> Sorry if it seems like I'm being argumentative here...I'm not intending to be.  I just care a lot about organizing RSpec's code in such a way that for optional features like the profiling, users only pay the runtime cost of that extra computation if they've opted-in to use the feature (kinda like how you care so much about stdlib usage!).  I think such an approach is essential to keeping RSpec fast even as we add and refine features.

I care too which is why it seems like we're being argumentative ;)

> Maybe what I'm asking for isn't easily possible for a reason I'm not yet seeing.  But I'd like to at least understand why if that's the case.

Comment (myronmarston) [inline lib/rspec/core.rb]: > Theres no two way communication required as long as you're ok with:

 > ```Ruby
config.profile_examples = true
config.profile_examples = false
```

> Not actually removing the profiler or the profile formatter.

I'm 100% OK with this.  In fact, I thought it was so obvious that I didn't think to mention it!

If the user toggles `profile_examples` from `true` to `false`, the correct user-facing behavior will still happen because we only notify the `dump_profile` event if the flag is set to `true`.  They'll still be paying the run-time cost of the profile stuff being tracked in memory but not output, but that's not a big deal, IMO.  The important things are that if `profile_examples` is set to true we produce the profile output and that if it is never set to `true` we don't waste time and resources loading and using the profiling logic.

**Comment by JonRowe** [comment]
I updated this some more, I'm not so convinced about the last commit but all the other stuff was good feedback :)

**Comment by myronmarston** [comment]
LGTM.  The build is failing due to rspec-expecations failures on 1.8.7, which I've attempted to address in rspec/rspec-expectations#798.

We can kick the build once that is merged but I think I'd rather just merge this w/o rebuilding since I referenced the failing build in the commit message of that other PR.

**Comment by JonRowe** [comment]
Turns out this needs a rebase anyway, and I've merged rspec/rspec-expectations#798 so lets give this us a go hey

**Comment by JonRowe** [comment]
This is still failing on 1.8.7 when combined with rspec-expectations despite the merging of the other PR, want to take a look @myronmarston?

**Comment by myronmarston** [comment]
> This is still failing on 1.8.7 when combined with rspec-expectations despite the merging of the other PR, want to take a look @myronmarston?

Done: rspec/rspec-mocks#963.
